### ISSUE TITLE
Pong DQN: flat learning curve at Adam 2.5e-4 — adopt Atari-standard Adam LR (6.25e-5) + LEARNING_RATE knob

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/thrust#342]
## Evidence (run 1, alc-2 RTX 4090, 2026-07-08)

First operator run of `train_pong_dqn` (#329/PR #341, 5M-wrapper-step budget) was stopped at the pre-declared no-learning decision point:

- **2.39M wrapper steps (≈9.56M raw frames), 2,809 episodes, ~6.2h wall-clock at ~103–130 wrapper steps/s.**
- `avg(last≤100)` **never exceeded −20.27** (achieved early, during the high-ε random phase) and sat at **−20.8 ± 0.1 from step ~1M onward** (ε floor). The greedy policy is marginally *worse* than random — the curve is flat with zero upward trend over 1.4M post-floor steps.
- Artifacts preserved on alc-2 under `~/pong_dqn_run1_lr2.5e-4/` (curve CSV, full log, 4 checkpoints at 500k intervals).

Under the Machado sticky-actions protocol some delay is expected, but published sticky-action DQN baselines (e.g. Dopamine) show visible improvement well before 10M raw frames on Pong. Flat-at-floor with greedy-worse-than-random through 9.5M raw frames indicates a training defect, not patience.

## Diagnosis

The example pairs **Adam with 2.5e-4** — the Mnih 2015 rate, but that rate was tuned for **RMSProp**. The Atari-DQN community standard for Adam is **6.25e-5** (Rainbow, Dopamine defaults; 4× lower). The trainer itself is healthy (Huber/Smooth-L1 loss, max_grad_norm 10, hard target sync 10k — verified during PR #341 review), which is consistent with "no divergence crash, but Q-updates too hot to converge."

The hyperparameter table in the example (PR #341) even records the mismatch: "| `learning_rate` | 2.5e-4 (RMSProp) | 2.5e-4 (Adam) |" — the Adam column should have been adapted, not copied.

## Proposed fix

- [ ] `examples/games/atari/train_pong_dqn.rs`: change the default `learning_rate` to **6.25e-5** and add a `LEARNING_RATE` env override (fail-loud parse, consistent with the other knobs).
- [ ] Update the module-doc hyperparameter table (Mnih RMSProp 2.5e-4 → ours Adam 6.25e-5, with the Rainbow/Dopamine citation) and the same table in `docs/PONG_DQN_RUNBOOK.md`.
- [ ] Runbook: add a short "run 1 negative result" note (flat at −20.8 through 9.5M raw frames at Adam 2.5e-4) so the next operator understands why the default changed — the honest-reporting protocol from #298 applies to failures too.
- [ ] Liftoff expectations: align the runbook's guidance with sticky-actions baselines (improvement typically visible 5–15M raw frames; crossing zero may take substantially longer than classic no-sticky DQN curves).

## References

#329 / PR #341 (the example), #306 (epic), Rainbow (Hessel et al. 2018) & Dopamine DQN defaults (Adam 6.25e-5), Machado et al. 2018 (sticky actions).

**Comment by rjwalters** [comment]
## Curator Enhancement

Curating this issue per user request. The issue is already well-evidenced and the fix plan is sound. This is a light-touch enrichment pass: code verification, `env_f64` prescription, tightened acceptance criteria, and a smoke-verification note.

---

### Code Verification Against origin/main

All findings confirmed against `examples/games/atari/train_pong_dqn.rs` and `docs/PONG_DQN_RUNBOOK.md` on `origin/main`.

**The `.learning_rate(2.5e-4)` call** — confirmed at line 241:
```rust
let config = DQNConfig::new()
    .learning_rate(2.5e-4)   // ← change to 6.25e-5
```
`DQNConfig.learning_rate` is typed `f64` (confirmed in `src/train/dqn/config.rs` line 18), so the new `env_f64` helper must parse into `f64`.

**The `env_usize` pattern to mirror** — the existing helper (line 405) is:
- Unset → silent fallback to default
- Present but unparseable → `anyhow!("{key}={s:?} is not a valid usize: {e}")` (fail-loud)

The new `env_f64` helper must follow identical semantics, substituting `f64` for `usize` and adding a positive-value guard (an operator entering `LEARNING_RATE=0` or a negative value should get a clear error, not a silently broken run). Suggested:
```rust
fn env_f64(key: &str, default: f64) -> Result<f64> {
    match std::env::var(key) {
        Ok(s) => {
            let v = s.parse::<f64>()
                .map_err(|e| anyhow::anyhow!("{key}={s:?} is not a valid f64: {e}"))?;
            if !v.is_finite() || v <= 0.0 {
                anyhow::bail!("{key}={s:?} must be a finite positive f64");
            }
            Ok(v)
        }
        Err(_) => Ok(default),
    }
}
```
The positive-finite guard is appropriate for a learning rate; other future `env_f64` callers (e.g. epsilon) that legitimately take zero should pass their own guard or accept the raw parsed value — document this in the function's doc comment.

**The `DEFAULT_LEARNING_RATE` constant** — the file declares other defaults as named module-level constants (`DEFAULT_TIMESTEPS`, `DEFAULT_BUFFER_CAPACITY`, `DEFAULT_MIN_BUFFER_SIZE`, `DEFAULT_LOG_INTERVAL`). Add:
```rust
/// Atari-standard Adam LR (Rainbow / Dopamine defaults). 4× lower than the
/// RMSProp 2.5e-4 from Mnih 2015; see issue #342.
const DEFAULT_LEARNING_RATE: f64 = 6.25e-5;
```
Then the call site becomes:
```rust
let learning_rate = env_f64("LEARNING_RATE", DEFAULT_LEARNING_RATE)?;
// ...
let config = DQNConfig::new()
    .learning_rate(learning_rate)
```
And `BurnOptimizer::new(inner_opt, config.learning_rate)` at line 254 naturally picks it up with no further change.

**The two hyperparameter tables** — both confirmed stale with `2.5e-4 (Adam)`:
- Module doc table: line 77 of `train_pong_dqn.rs`
- Runbook table: `docs/PONG_DQN_RUNBOOK.md` (the `| \`learning_rate\` |` row)

The citation format used elsewhere in the module doc is inline parenthetical: `(RMSProp)`, `(Adam)`. The updated cell should read `6.25e-5 (Adam; Rainbow/Dopamine)` to match the existing style and preserve the citation.

**The env-var knobs table in the module doc** (lines under `# Env-var knobs`) does not currently list `LEARNING_RATE` — add it:
```
- `LEARNING_RATE` (default `6.25e-5`) — Adam learning rate (Atari standard; Rainbow/Dopamine).
```

**The runbook env-var table** in `docs/PONG_DQN_RUNBOOK.md` likewise needs a `LEARNING_RATE` row added after `LOG_INTERVAL`.

---

### Scope Guard

**In scope (this issue only):**
- `examples/games/atari/train_pong_dqn.rs` — default LR change, `DEFAULT_LEARNING_RATE` const, `env_f64` helper, `LEARNING_RATE` knob wired in, module-doc table + env-var list updated
- `docs/PONG_DQN_RUNBOOK.md` — hyperparameter table updated, env-var table updated, run-1 negative result note added

**Out of scope (resist temptation):**
- `src/` changes of any kind — `DQNConfig`'s `learning_rate` builder method signature is already `f64`, no library changes needed
- Other hyperparameter changes (`target_update_interval`, `batch_size`, epsilon schedule, buffer size) — run-2 must be a single-variable experiment isolating the LR change; co-varying any other parameter contaminates the comparison

---

### Run-1 Negative Result Note (runbook placement)

Add a new subsection before Phase 0 (or as the opening of a "Experiment Log" section) in `docs/PONG_DQN_RUNBOOK.md`:

```markdown
## Experiment log

### Run 1 — Adam 2.5e-4 (negative result, 2026-07-08)

Stopped at the pre-declared no-learning decision point: **9.56M raw frames
(2.39M wrapper steps, ~6.2 h on alc-2 RTX 4090)**. `avg(last≤100)` never
exceeded −20.27 and sat at −20.8 ± 0.1 from ~1M wrapper steps onward (ε
floor). Greedy policy marginally worse than random — flat curve with zero
upward trend over 1.4M post-floor steps.

**Diagnosis:** Adam 2.5e-4 is the Mnih 2015 RMSProp rate; the Atari-standard
Adam rate is 6.25e-5 (Rainbow, Dopamine). The hyperparameter table's Adam
column was copied from the RMSProp column without adaptation (see issue #342).

**Resolution:** Default changed to 6.25e-5 in this PR. See run 2 below for
the corrected-LR result. Artifacts preserved on alc-2 under
`~/pong_dqn_run1_lr2.5e-4/`.
```

A "Run 2" subsection stub should be left as a placeholder (to be filled in post-run) so the operator knows where to append results.

---

### Liftoff / Baseline Alignment

The runbook's success criterion section says "Pong typically crosses zero well within ~10M raw frames." Under sticky actions (p=0.25) this is optimistic — published sticky-action baselines (Dopamine DQN, Machado et al. 2018) show crossing zero typically in the 5–15M raw frame range but sometimes later. The runbook should hedge this to: "crossing zero is typically seen within 10–20M raw frames under sticky-action settings; the 5M wrapper step (20M raw frame) budget is sized to cover this range." This is a clarification of existing text, not a new hyperparameter change.

---

## Acceptance Criteria

- [ ] `DEFAULT_LEARNING_RATE` constant defined as `6.25e-5` in `train_pong_dqn.rs`
- [ ] `env_f64` helper added with fail-loud semantics (unset → default, present-but-bad → hard error, non-finite or non-positive → hard error); doc comment notes the positive guard is LR-specific
- [ ] `LEARNING_RATE` env-var wired: `let learning_rate = env_f64("LEARNING_RATE", DEFAULT_LEARNING_RATE)?;` and passed to `.learning_rate(learning_rate)` and `BurnOptimizer::new(inner_opt, config.learning_rate)`
- [ ] Module-doc hyperparameter table row updated: `2.5e-4 (Adam)` → `6.25e-5 (Adam; Rainbow/Dopamine)`
- [ ] Module-doc env-var list gains `LEARNING_RATE` entry
- [ ] `docs/PONG_DQN_RUNBOOK.md` hyperparameter table row updated identically
- [ ] `docs/PONG_DQN_RUNBOOK.md` env-var table gains `LEARNING_RATE` row
- [ ] `docs/PONG_DQN_RUNBOOK.md` gains run-1 negative result note (experiment log or equivalent section) with the key data: 9.56M raw frames, −20.8 floor, diagnosis, alc-2 artifact path
- [ ] Runbook "crosses zero" baseline language hedged to 10–20M raw frame range under sticky actions
- [ ] No `src/` files modified; no other hyperparameter values changed

---

## Test Plan

- [ ] **Bad-value error:** `LEARNING_RATE=notanumber cargo run ... --example train_pong_dqn` exits immediately with a clear parse error naming `LEARNING_RATE`
- [ ] **Non-positive error:** `LEARNING_RATE=0 ...` and `LEARNING_RATE=-1e-4 ...` both produce a clear error (not a silent hang or NaN loss)
- [ ] **Unset defaults:** with `LEARNING_RATE` unset, the binary logs `learning_rate = 0.0000625` (or equivalent) in its startup banner and proceeds normally
- [ ] **Valid override applied:** `LEARNING_RATE=1e-4 ...` logs `learning_rate = 0.0001` in the startup banner and the value flows through to `config.learning_rate` (confirm via the startup log line, not just compilation)
- [ ] **Smoke run passes:** standard smoke verification — scratch venv, `pip install ale-py==0.12`, tiny budget (`TOTAL_TIMESTEPS=400 MIN_BUFFER_SIZE=100 BUFFER_CAPACITY=400`), CPU backend, confirms full lifecycle runs to completion with updated LR default
- [ ] **CI gates clean:** `cargo test --features "training,env-atari"` passes; `cargo clippy` produces no new warnings; `cargo fmt --check` clean

---

## Affected Files

- `examples/games/atari/train_pong_dqn.rs` — add `DEFAULT_LEARNING_RATE` const, add `env_f64` helper, wire `LEARNING_RATE` knob, update module-doc hyperparameter table and env-var list
- `docs/PONG_DQN_RUNBOOK.md` — update hyperparameter table, add `LEARNING_RATE` to env-var table, add run-1 negative result note, hedge liftoff baseline language

Both files confirmed present on `origin/main`. No `src/` files in scope.
